src/Microsoft.AspNet.WebHooks.Receivers/Extensions/WebHookHandlerContextExtensions.cs

```diff
@@ -30,16 +30,34 @@ public static T GetDataOrDefault<T>(this WebHookHandlerContext context)
                 return default(T);
             }
 
-            if (context.Data is JToken && !typeof(JToken).IsAssignableFrom(typeof(T)))
+            // IsAssignableFrom(...) check looks odd because return statement effectively assigns from Data to T.
+            // However, this check avoids useless "conversions" e.g. from a JObject to a JObject without attempting
+            // impossible conversions e.g. from a JObject to a JArray. Json.NET does not support semantically-invalid
+            // conversions between JToken subtypes.
+            //
+            // !typeof(T).IsAssignableFrom(context.Data.GetType()) may be more precise but is less efficient. That
+            // check would not change the (null) outcome in the JObject to JArray case, just adds a first-chance
+            // Exception (because the code would attempt the impossible conversion). About the only new cases it
+            // handles with a cast instead of ToObject<T>() are extreme corner cases such as when T is
+            // IDictionary<string, JToken> or another interface the current Data (e.g. JObject) may implement. Even
+            // then, Json.NET can usually perform an explicit conversion.
+            if (context.Data is JToken token && !typeof(JToken).IsAssignableFrom(typeof(T)))
             {
+                // context.Data is a JToken and the caller isn't just asking for (say) the current JObject.
+                // Do explicit conversion.
```

Comment:
So there's this huge comment here explaining the design tradeoffs but what does this code **mean**? Does it basically say that this block handles the case where T is one of the JLinq types?

Reply:
The block handles the case where `Data` is one of the JLinq types and `T` is *not* e.g. `Data` is a `JObject` and `T` is something like `AzureAlertNotification` (one of the strong types for request content). I'll add this info before the design "discussion".